### PR TITLE
Add header metas publish command

### PR DESCRIPTION
**Pull request by pxpm**
### BEFORE - What was wrong? What was happening before this PR?

It started with the idea of adding the capability to a Backpack panel to load on mobile devices as a "native app" when bookmarked on homepage in https://github.com/Laravel-Backpack/CRUD/pull/3318

While doing some information gathering to take the decision to do something about that PR, I found that by default Backpack does not even has a simple favicon meta 😞 

That needed to change 👍 

### AFTER - What is happening after this PR?

Developers can now publish a `header_metas` file along with some Backpack images as placeholders so that their apps have a favicon, display as an app in mobile, have colored navigation bar when displayed.

image
image
image

### How did you achieve that, in technical terms?

Added a command: `backpack:publish-header-metas` that will ask some questions to the developer, replace the stubs with the provided values and publish them. 

image

### Is it a breaking change?

NO, this needs https://github.com/Laravel-Backpack/theme-tabler/pull/185

### CODE REVIEW

File: src/app/Console/Commands/PublishHeaderMetas.php [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * Execute the console command.
+     */
+    public function handle()
+    {
+        $appName = $this->ask('Whats the application name ?', 'Backpack');
+        $appUrl = $this->ask('Whats the application url ?', config('app.url'));
+        $backpackPrefix = config('backpack.base.route_prefix');
+        $appColor = $this->ask('Whats the application color ?', '#605ca8');
```
Comment (tabacitu) [inline src/app/Console/Commands/PublishHeaderMetas.php:33]: I don't think making the app color purple is good for most projects. Personally I think that should be something muted, that stands out of the way. Here's what we use for Backpack, I think it's a better default:

```suggestion
        $appColor = $this->ask('What is the application color?', '#161c2d');
```

image
 versus the bright purple (Backpack's primary color) which looks like this
image
or the muted purple that was here before:
image

File: src/resources/stubs/header_metas.stub [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<meta name="apple-mobile-web-app-capable" content="yes" />
```
Comment (tabacitu) [inline src/resources/stubs/header_metas.stub:1]: I've just generated our icons again using https://realfavicongenerator.net which is my favorite, and it has a few stuff in addition to what is here. Namely the Windows stuff. Not sure that's really relevant any more in 2024... you should know better Pedro.

Here's the whole zip with it - I've chosed `/admin` as my prefix there, and `Backpack` as my app title:

[favicon_package_v0.16 (4).zip](https://github.com/user-attachments/files/16157735/favicon_package_v0.16.4.zip)

File: src/app/Console/Commands/PublishHeaderMetas.php [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+     * @var string
+     */
+    protected $signature = 'backpack:publish-header-metas';
+
+    /**
+     * The console command description.
+     *
+     * @var string
+     */
+    protected $description = 'Publishes the header metas and assets.';
```
Comment (tabacitu) [inline src/app/Console/Commands/PublishHeaderMetas.php:23]:
```suggestion
    protected $description = 'Publishes the header metas and favicon assets.';
```

**Review by tabacitu — CHANGES_REQUESTED** [review]
Tested, works as expected on local! 

I like this a lot, I like how it's not actually anything to maintain, we just help them generate some useful code and files. Good job Pedro!

My only real gripe is with the absolute paths (see my comments) but I've taken the time to scrutinize everything, so I don't have to go through it again 😀

Take a look at my comments please and if you agree it's feasible... run with it!

File: src/app/Console/Commands/PublishHeaderMetas.php [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+
+    /**
+     * Execute the console command.
+     */
+    public function handle()
+    {
+        $appName = $this->ask('Whats the application name ?', 'Backpack');
+        $appUrl = $this->ask('Whats the application url ?', config('app.url'));
```
[the remark this answers is not in the document]
Comment (pxpm) [reply src/app/Console/Commands/PublishHeaderMetas.php:31]: Nice catch! 👍 

Thanks, i've just refactored it like you proposed.

File: src/app/Console/Commands/PublishHeaderMetas.php [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+     * @var string
+     */
+    protected $description = 'Publishes the header metas and assets.';
+
+    /**
+     * Execute the console command.
+     */
+    public function handle()
+    {
+        $appName = $this->ask('Whats the application name ?', 'Backpack');
```
[the remark this answers is not in the document]
Comment (pxpm) [reply src/app/Console/Commands/PublishHeaderMetas.php:30]: Did it 👍 Nice suggestion and thinking!

**Comment by dimer47** [comment]
This is a fantastic feature @pxpm ! But there is a slight incompatibility between the commands `php artisan backpack:publish-header-metas` and `php artisan backpack:install -q -n`.

When using the `php artisan backpack:publish-header-metas` command, it prompts for the installation location of the assets. However, by default, the `php artisan backpack:install -q -n` command installs everything in the `public/` directory.

The `BackpackServiceProvider.php` file located in `vendor/backpack/crud/src/BackpackServiceProvider.php` contains the following code:
```php
public function publishFiles()
{
    ...
    $backpack_public_assets = [__DIR__.'/public' => public_path()];
    ...
}
```
Should the assets really be deployed by default during the installation? Or wouldn't it be better to prompt for the installation location of the assets during the installation, with the default value set to `/public`?

May I create a PR to implement this request, or do you have a different perspective on the matter?

**Comment by pxpm** [comment]
Hey @dimer47 nice catch! Completely missed that. 😞 

I think the correct thing to do is removing that `$backpack_public_assets` from the publish command as there is nothing in that folder that we would like to publish. 

If you have time to create the PR, please go ahead, I will merge and tag a new version right way. 👍

**Comment by dimer47** [comment]
Hey @pxpm, sorry for the delay! 😅

I've gone ahead and created the PR as requested. I've removed the $backpack_public_assets from the publishFiles method, so no assets will be published during the backpack:install command anymore. This ensures that asset deployment is only triggered by the backpack:publish-header-metas command, just as we discussed.

Merge Request : https://github.com/Laravel-Backpack/CRUD/pull/5619

Thank you for your patience. Looking forward to the merge. 😊

**Pull request merged**